src/transformer/mockDefiner/cache/factoryDeclarationListCache.ts

```diff
@@ -0,0 +1,44 @@
+import * as ts from 'typescript';
+
+export interface DeclarationCache {
+    declarations: ts.Declaration[];
+    key: string;
+}
+
+export class FactoryDeclarationListCache {
+    private _cache: DeclarationCache[];
+
+    constructor() {
+        this._cache = [];
+    }
+
+    public set(declarations: ts.Declaration[], key: string): void {
+        this._cache.push({
+            declarations,
+            key,
+        });
+    }
+
+    public get(declarations: ts.Declaration[]): string {
+        return this._find(declarations).key;
+    }
+
+    public has(declarations: ts.Declaration[]): boolean {
+        return !!this._find(declarations);
+    }
+
+    private _find(declarations: ts.Declaration[]): DeclarationCache {
+        return this._cache.find((intersection: DeclarationCache) => {
+            const declarationsCopy: ts.Declaration[] = [...declarations];
```

Comment:
Can you tell me what this algorithm is doing? It seems like you're checking if `intersection.declarations` contains `declarations`. If I'm correct can you explain the reasoning behind this approach?

Reply:
Now that I'm reading it I'm finding that ArrayA is a subset of ArrayB. I should add the length check at the beginning.
Considering that I will fix with that check I am doing this to make sure that the array have the same elements.
If I don't create the copy and just check of the IndexOf this scenario will result true even if is false

arrayA = ['a', 'b']
arrayB = ['a', 'a']

This is what I'm doing

- Create a copy of arrayA
- find for each Element of ArrayB the element on ArrayA
- remove the element from the ArrayA
- If the arrayA is Empty it means is good

Reply:
I would write this function somewhere as a utility so you have a place where to write unit tests for it

Reply:
Done. It's actually good having a setup for some unit test.

Reply:
Basically, the last test would not pass using https://jsperf.com/pmyl-compare-array iterative example because all the element in the first array are present in the second array as first index

Reply:
I've done some more alternatives in https://jsperf.com/pmyl-compare-array/6
I don't really know if it's worth it to change it, I can approve the pr now, let's keep in mind that this could be quicker with some investigation.
I wouldn't do it now since I expect, after reading the new issues, some of this code to change at some point.